File: Chapter 10/JS/section_10_3_resilience_decorators/resilient/infrastructure/adapters/flakyPaymentsPaymentAdapter.js
import { from, timer, throwError } from 'rxjs';
import { retry, mergeMap, tap } from 'rxjs/operators';
import { PaymentGateway } from '../../core/ports/PaymentGateway.js';

/**
 * THE INFRASTRUCTURE ADAPTER (The Implementation):
 * * @description
 * This class encapsulates the Physical Resource Policy using RxJS. 
 * It transforms raw network instability into a predictable, shielded stream.
 *
 * ARCHITECTURAL CRITIQUE:
 * 1. POLICY AS DOCUMENTATION: By maintaining these constants, the Adapter 
 * serves as the living documentation for our vendor SLA.
 * * 2. RECOVERY BOUNDARY: The use of 'BACKOFF_FACTOR' and 'MAX_RETRIES' 
 * ensures we don't accidentally DOS our own vendor during a recovery spike.
 * * 3. FAIL-FAST ON EXHAUSTION: We have optimized the retry logic to ensure 
 * that once the MAX_RETRIES is reached, we throw the error immediately 
 * instead of backing off one last time. This prevents "Temporal Leakage" 
 * where the system hangs for no reason before failing over to the queue.
 */
export class FlakyPaymentsPaymentAdapter extends PaymentGateway {
    static CONNECT_TIMEOUT_MS = 2000;
    static READ_TIMEOUT_MS = 8000;
    static MAX_RETRIES = 5;
    static MIN_TIMEOUT_MS = 2000;
    static MAX_TIMEOUT_MS = 10000;
    static BACKOFF_FACTOR = 2;

    constructor(baseUrl = "https://api.flakypayments.com") {
        super();
        this.baseUrl = baseUrl;
    }

    /**
     * @override
     * @returns {import('rxjs').Observable<boolean>}
     */
    charge(amount, orderId, idempotencyKey) {
        return from(this._performRawRequest(amount, orderId, idempotencyKey)).pipe(
            retry({
                count: FlakyPaymentsPaymentAdapter.MAX_RETRIES,
                delay: (error, retryCount) => {
                    // LOG THE FAILURE REGARDLESS OF THE ATTEMPT COUNT
                    console.warn(`      [Retry Shield] Attempt ${retryCount} failed: ${error.message}.`);

                    // IF WE HAVE EXHAUSTED RETRIES: 
                    // Terminate the stream immediately so the Orchestrator can pivot.
                    if (retryCount >= FlakyPaymentsPaymentAdapter.MAX_RETRIES) {
                        console.error(`      [Retry Shield] MAX_RETRIES (${FlakyPaymentsPaymentAdapter.MAX_RETRIES}) reached. Exhausted.`);
                        return throwError(() => error);
                    }

                    const backoff = FlakyPaymentsPaymentAdapter.MIN_TIMEOUT_MS * Math.pow(FlakyPaymentsPaymentAdapter.BACKOFF_FACTOR, retryCount - 1);
                    const delayTime = Math.min(backoff, FlakyPaymentsPaymentAdapter.MAX_TIMEOUT_MS);
                    
                    console.warn(`      [Retry Shield] Backing off ${delayTime}ms...`);
                    
                    return timer(delayTime);
                }
            }),
            tap(() => console.log(`      [FlakyPayments Adapter] SUCCESS: Order ${orderId} charged.`))
        );
    }

    async _performRawRequest(amount, order_id, key) {
        // Simulated failure to test the shield
        throw new Error("FlakyPayments API: Gateway Timeout (504)");
    }
}

File: Chapter 10/Python/section_10_3_resilience_decorators/resilient/infrastructure/adapters/flaky_payments_payment_adapter.py
import requests
import math
from tenacity import retry, stop_after_attempt, wait_exponential, after_log, before_sleep_log
import logging
from ...core.ports.payment_gateway import PaymentGateway

# Configure a minimalist logger for the Retry Shield
logger = logging.getLogger(__name__)

class FlakyPaymentsPaymentAdapter(PaymentGateway):
    """
    THE INFRASTRUCTURE ADAPTER (The Implementation):
    
    ARCHITECTURAL CRITIQUE:
    1. OBSERVABILITY: We log every failure and the specific 
       backoff duration. This prevents "Silent Failures" 
       and makes the system's struggle visible to the operator.
    * 2. FAIL-FAST: The retry policy is encapsulated here. When 
       MAX_RETRY_ATTEMPTS is reached, tenacity reraises the last error 
       instantly, allowing the Orchestrator to pivot to Plan B.
    """

    # --- THE PHYSICAL POLICY CONSTANTS (The SLA) ---
    CONNECT_TIMEOUT_SEC = 2
    READ_TIMEOUT_SEC = 8
    MAX_RETRY_ATTEMPTS = 5
    BACKOFF_MIN_SEC = 2
    BACKOFF_MAX_SEC = 10
    BACKOFF_FACTOR = 2

    def __init__(self, base_url: str = "https://api.flakypayments.com"):
        self.base_url = base_url

    def _log_retry(self, retry_state):
        """Custom callback to match the JS 'Senior Architect' log style."""
        attempt_number = retry_state.attempt_number
        error_msg = retry_state.outcome.exception()
        
        print(f"      [Retry Shield] Attempt {attempt_number} failed: {error_msg}.")
        
        if attempt_number >= self.MAX_RETRY_ATTEMPTS:
            print(f"      [Retry Shield] MAX_RETRIES ({self.MAX_RETRY_ATTEMPTS}) reached. Exhausted.")
        else:
            # Calculate next backoff for logging parity
            delay = min(self.BACKOFF_MIN_SEC * (self.BACKOFF_FACTOR ** (attempt_number - 1)), self.BACKOFF_MAX_SEC)
            print(f"      [Retry Shield] Backing off {int(delay * 1000)}ms...")

    # THE RESILIENCE SHIELD
    @retry(
        stop=stop_after_attempt(MAX_RETRY_ATTEMPTS),
        wait=wait_exponential(multiplier=1, min=BACKOFF_MIN_SEC, max=BACKOFF_MAX_SEC),
        # Attach our custom logging callback
        after=lambda retry_state: None, # Placeholder for standard tenacity logging
        before_sleep=lambda retry_state: FlakyPaymentsPaymentAdapter._log_retry_callback(retry_state),
        reraise=True
    )
    def charge(self, amount: float, order_id: str, idempotency_key: str) -> bool:
        print(f"      [Infrastructure Adapter] Attempting FlakyPayments Charge for {order_id}...")
        
        # SIMULATION: To trigger the shield as seen in your JS examples
        # raise requests.exceptions.RequestException("FlakyPayments API: Gateway Timeout (504)")
        
        response = requests.post(
            f"{self.base_url}/charge",
            json={"amount": amount, "order_id": order_id},
            headers={"Idempotency-Key": idempotency_key},
            timeout=(self.CONNECT_TIMEOUT_SEC, self.READ_TIMEOUT_SEC)
        )
        response.raise_for_status()
        return True

    @staticmethod
    def _log_retry_callback(retry_state):
        """Helper to route tenacity state to our formatted print statements."""
        # Accessing constants via the class reference
        adapter = FlakyPaymentsPaymentAdapter
        attempt = retry_state.attempt_number
        ex = retry_state.outcome.exception()
        
        print(f"      [Retry Shield] Attempt {attempt} failed: {ex}.")
        
        if attempt >= adapter.MAX_RETRY_ATTEMPTS:
            print(f"      [Retry Shield] MAX_RETRIES ({adapter.MAX_RETRY_ATTEMPTS}) reached. Exhausted.")
        else:
            # Tenacity 'next_action' provides the sleep duration
            delay_ms = int(retry_state.next_action.sleep * 1000)
            print(f"      [Retry Shield] Backing off {delay_ms}ms...")